### PR TITLE
feat(sdk): Add an action parameter to the OIDC account URL.

### PR DESCRIPTION
**Pull request by pixlwave**
This PR adds the `action` and `device_id` account management URL parameters to OIDC and the FFI as detailed here:

https://github.com/sandhose/matrix-doc/blob/msc/sandhose/oidc-discovery/proposals/2965-oidc-discovery.md#account-management-url-parameters

Unsure if its possible to test these currently, I'm hoping @bnjbvr will add enough mocks to make that possible. For now I have checked they're all working against a server.

Part of https://github.com/vector-im/element-internal/issues/487

### CODE REVIEW

File: bindings/matrix-sdk-ffi/src/client.rs [hunk 1]
```diff
@@ -430,8 +430,13 @@ impl Client {
         })
     }
 
-    pub fn account_url(&self) -> Option<String> {
-        self.inner.oidc().account_management_url().ok().flatten().map(|url| url.to_string())
+    pub fn account_url(&self, action: Option<AccountManagementAction>) -> Option<String> {
+        self.inner
+            .oidc()
+            .account_management_url(action.map(Into::into))
+            .ok()
```
Comment (bnjbvr) [inline bindings/matrix-sdk-ffi/src/client.rs:437]: Instead of this `ok()` call, can you use a match, and log the error in the `Err` case please?

**Review by bnjbvr — APPROVED** [review]

**Comment by pixlwave** [comment]
> Can we also please put the bindings change into a separate commit.

> it's a good practice to keep every commit compiling, even if it is clippy that's failing.

I think these might be opposing each other? If I put the bindings change into a separate commit, the bindings will fail to compile because the call to `account_management_url` will be missing the parameter?

**Comment by poljar** [comment]
> > Can we also please put the bindings change into a separate commit.
> 
> > it's a good practice to keep every commit compiling, even if it is clippy that's failing.
> 
> I think these might be opposing each other? If I put the bindings change into a separate commit, the bindings will fail to compile because the call to `account_management_url` will be missing the parameter?

Well yes, but the new parameter is an `Option` so you can pass in `None` in the initial commit to keep things compiling and introduce the change to the public API of the bindings in the separate commit.

File: crates/matrix-sdk/src/oidc/mod.rs [hunk 2]
```diff
@@ -1225,3 +1267,110 @@ fn hash<T: Hash>(x: &T) -> u64 {
...
+    use crate::{
+        oidc::{
+            FullSession, OidcAccountManagementAction, RegisteredClientData, SessionTokens,
+            UserSession,
+        },
+        ClientBuilder,
+    };
+
+    #[async_test]
+    async fn test_account_management_url() {
```
Comment (pixlwave) [inline crates/matrix-sdk/src/oidc/mod.rs:1294]: I'm totally assuming that the below doesn't make an requests to `example.com`?

Comment (bnjbvr) [reply crates/matrix-sdk/src/oidc/mod.rs:1294]: No indeed.

**Review by bnjbvr — APPROVED** [review]
Noice!

**Comment by pixlwave** [comment]
> introduce the change to the public API of the bindings in the separate commit.

Right, now I understand the reasoning for splitting it up 👍

**Review by poljar — APPROVED** [review]
Thanks. Looking good now.

**Pull request merged**